### PR TITLE
Replace make_typecast by typecast_exprt or typecast_exprt::conditional_cast [blocks: #3800]

### PR DESCRIPTION
**Pull request by tautschnig**
make_typecast is deprecated.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: src/goto-programs/builtin_functions.cpp [hunk 1]
```diff
@@ -513,8 +511,8 @@ void goto_convertt::do_cpp_new(
     new_call.add_source_location()=rhs.source_location();
 
     for(std::size_t i=0; i<code_type.parameters().size(); i++)
-      if(new_call.arguments()[i].type()!=code_type.parameters()[i].type())
-        new_call.arguments()[i].make_typecast(code_type.parameters()[i].type());
+      new_call.arguments()[i] = typecast_exprt::conditional_cast(
```
Comment (smowton) [inline src/goto-programs/builtin_functions.cpp:515]: Braces around multi-line `for` please

File: src/goto-programs/builtin_functions.cpp [hunk 2]
```diff
@@ -1530,10 +1533,12 @@ void goto_convertt::do_function_call_symbol(
     if(arguments[0].type()!=arguments[1].type())
     {
       if(use_double)
-        new_arguments[1].make_typecast(arguments[0].type());
+        new_arguments[1] =
```
Comment (smowton) [inline src/goto-programs/builtin_functions.cpp:1539]: braces

File: src/goto-programs/goto_convert.cpp [hunk 3]
```diff
@@ -718,16 +718,14 @@ void goto_convertt::convert_assign(
 
     if(lhs.id() == ID_typecast)
     {
-      DATA_INVARIANT(
-        lhs.operands().size() == 1, "Typecast must have one operand");
+      UNREACHABLE; // the original code here was broken, presumably it's never used
+      typecast_exprt tc = to_typecast_expr(lhs);
```
Comment (smowton) [inline src/goto-programs/goto_convert.cpp]: I think some compilers will give you a hard time for this unreachable code

File: src/goto-programs/pointer_arithmetic.cpp [hunk 4]
```diff
@@ -71,14 +71,8 @@ void pointer_arithmetict::add_to_offset(const exprt &src)
...
   else
-  {
-    exprt new_offset=plus_exprt(offset, src);
-
-    if(new_offset.op1().type()!=offset.type())
-      new_offset.op1().make_typecast(offset.type());
-
-    offset=new_offset;
-  }
+    offset =
```
Comment (smowton) [inline src/goto-programs/pointer_arithmetic.cpp:75]: braces

File: src/pointer-analysis/value_set.cpp [hunk 5]
```diff
@@ -1090,7 +1090,8 @@ void value_sett::get_reference_set_rec(
         {
           // adjust type?
           if(ns.follow(struct_op.type())!=final_object_type)
-            member_expr.compound().make_typecast(struct_op.type());
+            member_expr.compound() =
```
Comment (smowton) [inline src/pointer-analysis/value_set.cpp:1093]: braces

File: src/pointer-analysis/value_set_fi.cpp [hunk 6]
```diff
@@ -914,7 +914,8 @@ void value_set_fit::get_reference_set_sharing_rec(
 
         // adjust type?
         if(ns.follow(struct_op.type())!=ns.follow(object.type()))
-          member_expr.compound().make_typecast(struct_op.type());
+          member_expr.compound() =
```
Comment (smowton) [inline src/pointer-analysis/value_set_fi.cpp:918]: braces

File: src/pointer-analysis/value_set_fivr.cpp [hunk 7]
```diff
@@ -1027,7 +1027,8 @@ void value_set_fivrt::get_reference_set_sharing_rec(
 
         // adjust type?
         if(ns.follow(struct_op.type())!=ns.follow(object.type()))
-          member_expr.compound().make_typecast(struct_op.type());
+          member_expr.compound() =
```
Comment (smowton) [inline src/pointer-analysis/value_set_fivr.cpp:1031]: braces

File: src/pointer-analysis/value_set_fivrns.cpp [hunk 8]
```diff
@@ -720,7 +720,8 @@ void value_set_fivrnst::get_reference_set_rec(
 
         // adjust type?
         if(ns.follow(struct_op.type())!=ns.follow(object.type()))
-          member_expr.compound().make_typecast(struct_op.type());
+          member_expr.compound() =
```
Comment (smowton) [inline src/pointer-analysis/value_set_fivrns.cpp:724]: braces

**Review by smowton — APPROVED** [review]
Few places where there's now a multi-line statement after an `if` or `for` which I'd prefer to be braced, otherwise lgtm

File: src/goto-programs/goto_convert.cpp (hunk 3 above)
Comment (tautschnig) [reply src/goto-programs/goto_convert.cpp]: I have no idea how we would generate an expression that has a typecast on the left-hand side, but even if and when we do: at least goto-symex does handle that case, so we could safely leave it in the goto program. I am going to add a commit that simply removes this code.

File: src/goto-programs/builtin_functions.cpp (hunk 1 above)
Comment (tautschnig) [reply src/goto-programs/builtin_functions.cpp:515]: Done.

File: src/goto-programs/builtin_functions.cpp (hunk 2 above)
Comment (tautschnig) [reply src/goto-programs/builtin_functions.cpp:1539]: Done.

File: src/goto-programs/pointer_arithmetic.cpp (hunk 4 above)
Comment (tautschnig) [reply src/goto-programs/pointer_arithmetic.cpp:75]: Done.

File: src/pointer-analysis/value_set.cpp (hunk 5 above)
Comment (tautschnig) [reply src/pointer-analysis/value_set.cpp:1093]: Done.

File: src/pointer-analysis/value_set_fi.cpp (hunk 6 above)
Comment (tautschnig) [reply src/pointer-analysis/value_set_fi.cpp:918]: Done.

File: src/pointer-analysis/value_set_fivr.cpp (hunk 7 above)
Comment (tautschnig) [reply src/pointer-analysis/value_set_fivr.cpp:1031]: Done.

File: src/pointer-analysis/value_set_fivrns.cpp (hunk 8 above)
Comment (tautschnig) [reply src/pointer-analysis/value_set_fivrns.cpp:724]: Done.

File: src/goto-programs/goto_convert.cpp (hunk 3 above)
Comment (smowton) [reply src/goto-programs/goto_convert.cpp]: We definitely do sometimes produce `void** data = new void*[10]; ((Object**)data)[0] = new Object` in the Java frontend. Whether we should is another question!

Comment (tautschnig) [reply src/goto-programs/goto_convert.cpp]: Sure, and you can make something similar happen in C (`*(char*)&x = 42;`) - but there's an `index_exprt` on the left-hand side in your example and a `dereference_exprt` in my case. Neither has a typecast at the top level on the left-hand side.

Comment (smowton) [reply src/goto-programs/goto_convert.cpp]: True. In the past we would sometimes do `(Object*)var = new Object`, but AFAIK that's no longer generated by the frontend

Comment (tautschnig) [reply src/goto-programs/goto_convert.cpp]: I think even if that were still the case, goto-symex would happily take care of that.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 9ee63c48fd9774f0ef55e751812d56281a7dcf5b).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/99188565

File: src/goto-programs/goto_convert.cpp (hunk 3 above)
Comment (tautschnig) [reply src/goto-programs/goto_convert.cpp]: Just for the record, those changes have now been moved to #3770, this PR now exclusively deals with replacing `make_typecast`.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 86852099c4773eca7f5308e2f079b3db1737f254).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/99534366

**Review by peterschrammel — APPROVED** [review]

File: src/ansi-c/c_typecheck_expr.cpp [hunk 9]
```diff
@@ -2955,7 +2956,7 @@ void c_typecheck_baset::typecheck_expr_binary_arithmetic(exprt &expr)
     is_number(o_type1))
   {
     // convert op1 to the vector type
-    op1.make_typecast(o_type0);
+    op1 = typecast_exprt::conditional_cast(op1, o_type0);
     expr.type() = o_type0;
```
Comment (kroening) [inline src/ansi-c/c_typecheck_expr.cpp:2958]: A conditional_cast appears to be wrong here. When we are in this branch, the cast is always needed. Simply use the constructor.

File: src/ansi-c/c_typecheck_expr.cpp [hunk 10]
```diff
@@ -2964,7 +2965,7 @@ void c_typecheck_baset::typecheck_expr_binary_arithmetic(exprt &expr)
     is_number(o_type0))
   {
     // convert op0 to the vector type
-    op0.make_typecast(o_type1);
+    op0 = typecast_exprt::conditional_cast(op0, o_type1);
     expr.type() = o_type1;
```
Comment (kroening) [inline src/ansi-c/c_typecheck_expr.cpp:2967]: Same here.

File: src/cpp/cpp_typecheck_conversions.cpp [hunk 11]
```diff
@@ -220,24 +219,21 @@ bool cpp_typecheckt::standard_conversion_integral_promotion(
     std::size_t width=to_unsignedbv_type(expr.type()).get_width();
     if(width >= config.ansi_c.int_width)
       return false;
-    new_expr=expr;
     if(width==config.ansi_c.int_width)
       int_type.id(ID_unsignedbv);
-    new_expr.make_typecast(int_type);
+    new_expr = typecast_exprt::conditional_cast(expr, int_type);
     return true;
```
Comment (kroening) [inline src/cpp/cpp_typecheck_conversions.cpp:223]: Same here -- the cast will always happen.

File: src/cpp/cpp_typecheck_conversions.cpp [hunk 12]
```diff
@@ -271,8 +267,7 @@ bool cpp_typecheckt::standard_conversion_floating_point_promotion(
   c_qualifierst qual_from;
   qual_from.read(expr.type());
 
-  new_expr=expr;
-  new_expr.make_typecast(double_type());
+  new_expr = typecast_exprt::conditional_cast(expr, double_type());
   qual_from.write(new_expr.type());
```
Comment (kroening) [inline src/cpp/cpp_typecheck_conversions.cpp:269]: same here

File: src/goto-cc/linker_script_merge.cpp [hunk 13]
```diff
@@ -576,8 +576,8 @@ int linker_script_merget::ls_data2instructions(
         unsigned_int_type().get_width()),
       unsigned_int_type());
 
-    exprt rhs_tc(rhs);
-    rhs_tc.make_typecast(pointer_type(char_type()));
+    exprt rhs_tc =
+      typecast_exprt::conditional_cast(rhs, pointer_type(char_type()));
```
Comment (kroening) [inline src/goto-cc/linker_script_merge.cpp:580]: Same here

File: src/goto-programs/builtin_functions.cpp [hunk 14]
```diff
@@ -441,10 +441,8 @@ void goto_convertt::do_cpp_new(
 
   if(new_array)
   {
-    count=static_cast<const exprt &>(rhs.find(ID_size));
-
-    if(count.type()!=object_size.type())
-      count.make_typecast(object_size.type());
+    count = typecast_exprt(
+      static_cast<const exprt &>(rhs.find(ID_size)), object_size.type());
```
Comment (kroening) [inline src/goto-programs/builtin_functions.cpp:446]: But here it's the conditional_cast

File: src/pointer-analysis/value_set_dereference.cpp [hunk 15]
```diff
@@ -382,7 +382,7 @@ value_set_dereferencet::valuet value_set_dereferencet::build_reference_to(
       equal_exprt equality(pointer_expr, object_pointer);
 
       if(ns.follow(equality.lhs().type())!=ns.follow(equality.rhs().type()))
-        equality.lhs().make_typecast(equality.rhs().type());
+        equality.lhs() = typecast_exprt(equality.lhs(), equality.rhs().type());
```
Comment (kroening) [inline src/pointer-analysis/value_set_dereference.cpp]: Conditional cast!

File: src/pointer-analysis/value_set_dereference.cpp [hunk 16]
```diff
@@ -452,7 +452,7 @@ value_set_dereferencet::valuet value_set_dereferencet::build_reference_to(
       result.value=index_expr;
 
       if(ns.follow(result.value.type())!=ns.follow(dereference_type))
-        result.value.make_typecast(dereference_type);
+        result.value = typecast_exprt(result.value, dereference_type);
     }
```
Comment (kroening) [inline src/pointer-analysis/value_set_dereference.cpp:449]: Same here

File: src/pointer-analysis/value_set_dereference.cpp [hunk 17]
```diff
@@ -594,7 +594,7 @@ bool value_set_dereferencet::memory_model_bytes(
 
     // possibly need to convert
     if(!base_type_eq(result.type(), to_type, ns))
-      result.make_typecast(to_type);
+      result = typecast_exprt(result, to_type);
   }
```
Comment (kroening) [inline src/pointer-analysis/value_set_dereference.cpp:588]: Same here

**Review by kroening — APPROVED** [review]

File: src/ansi-c/c_typecheck_expr.cpp (hunk 9 above)
Comment (tautschnig) [reply src/ansi-c/c_typecheck_expr.cpp:2958]: Done.

File: src/ansi-c/c_typecheck_expr.cpp (hunk 10 above)
Comment (tautschnig) [reply src/ansi-c/c_typecheck_expr.cpp:2967]: Done.

File: src/cpp/cpp_typecheck_conversions.cpp (hunk 11 above)
Comment (tautschnig) [reply src/cpp/cpp_typecheck_conversions.cpp:223]: I don't think so - the line above the change would make the typecast a no-op.

File: src/cpp/cpp_typecheck_conversions.cpp (hunk 12 above)
Comment (tautschnig) [reply src/cpp/cpp_typecheck_conversions.cpp:269]: Done.

File: src/goto-cc/linker_script_merge.cpp (hunk 13 above)
Comment (tautschnig) [reply src/goto-cc/linker_script_merge.cpp:580]: Done.

File: src/goto-programs/builtin_functions.cpp (hunk 14 above)
Comment (tautschnig) [reply src/goto-programs/builtin_functions.cpp:446]: Done.

File: src/pointer-analysis/value_set_dereference.cpp (hunk 15 above)
Comment (tautschnig) [reply src/pointer-analysis/value_set_dereference.cpp]: Done (killing the ns.follow)

File: src/pointer-analysis/value_set_dereference.cpp (hunk 16 above)
Comment (tautschnig) [reply src/pointer-analysis/value_set_dereference.cpp:449]: Done (killing ns.follow)

File: src/pointer-analysis/value_set_dereference.cpp (hunk 17 above)
Comment (tautschnig) [reply src/pointer-analysis/value_set_dereference.cpp:588]: Done (killing this instance of `base_type_eq`)

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 76b2c26dec7d06ebeb0c35da12079b83bb7c49cf).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/100052504

File: src/cpp/cpp_typecheck_conversions.cpp (hunk 11 above)
Comment (kroening) [reply src/cpp/cpp_typecheck_conversions.cpp:223]: Unrelated to PR: the two lines above the change are dead code (note the branch directly before). There's probably a bug there.

Comment (tautschnig) [reply src/cpp/cpp_typecheck_conversions.cpp:223]: Oh, indeed. Aren't those two lines really just unnecessary, and we should do an unconditional cast as you suggested? To me it seems there no bug other than it being dead code.

Comment (kroening) [reply src/cpp/cpp_typecheck_conversions.cpp:223]: The bug is that an "unsigned int" will be promoted to "signed int", which is wrong.

Comment (tautschnig) [reply src/cpp/cpp_typecheck_conversions.cpp:223]: I don't think so, the `width >= config.ansi_c.int_width` would catch that?

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 9ece3a7aa97d5eb34ad78cbc119cb0eae553547c).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/100077874
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: f6948db50484afb698081b2f944547863b3f9b36).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/100080400

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: ebaaceabb6224d74a4356b88e7a509d1e7887bb6).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/100100359

**Pull request merged**